File: src/test/java/org/stellar/sdk/TOIDTest.java
package org.stellar.sdk;

import java.util.Arrays;
import java.util.List;
import org.junit.Assert;
import org.junit.Test;

public class TOIDTest {

  private static final long ledgerFirst = 1L << 32;
  private static final long txFirst = 1L << 12;
  private static final long opFirst = 1L;

  @Test
  public void testToInt64AndFromInt64() {
    List<TOID> toids =
        Arrays.asList(
            new TOID(0, 0, 1),
            new TOID(0, 0, 4095),
            new TOID(0, 1, 0),
            new TOID(0, 1048575, 0),
            new TOID(1, 0, 0),
            new TOID(2_147_483_647, 0, 0),
            new TOID(1, 1, 1),
            new TOID(1, 1, 0),
            new TOID(1, 0, 1),
            new TOID(1, 0, 0),
            new TOID(0, 1, 0),
            new TOID(0, 0, 1),
            new TOID(0, 0, 0),
            new TOID(2_147_483_647, 1_048_575, 4_095));
    List<Long> ids =
        Arrays.asList(
            1L,
            4095L,
            4096L,
            4294963200L,
            4294967296L,
            9223372032559808512L,
            ledgerFirst + txFirst + opFirst,
            ledgerFirst + txFirst,
            ledgerFirst + opFirst,
            ledgerFirst,
            txFirst,
            opFirst,
            0L,
            9223372036854775807L);
    for (int i = 0; i < toids.size(); i++) {
      TOID toid = toids.get(i);
      long id = ids.get(i);
      Assert.assertEquals(id, toid.toInt64());
      Assert.assertEquals(toid, TOID.fromInt64(id));
    }
  }

  @Test
  public void testLedgerRangeInclusive() {
    Assert.assertEquals(
        new TOID.TOIDRange(new TOID(0, 0, 0).toInt64(), new TOID(2, 0, 0).toInt64()),
        TOID.ledgerRangeInclusive(1, 1));
    Assert.assertEquals(
        new TOID.TOIDRange(new TOID(0, 0, 0).toInt64(), new TOID(3, 0, 0).toInt64()),
        TOID.ledgerRangeInclusive(1, 2));
    Assert.assertEquals(
        new TOID.TOIDRange(new TOID(2, 0, 0).toInt64(), new TOID(3, 0, 0).toInt64()),
        TOID.ledgerRangeInclusive(2, 2));
    Assert.assertEquals(
        new TOID.TOIDRange(new TOID(2, 0, 0).toInt64(), new TOID(4, 0, 0).toInt64()),
        TOID.ledgerRangeInclusive(2, 3));
  }

  @Test
  public void testIncrementOperationIndex() {
    TOID toid = new TOID(0, 0, 0);
    toid.incrementOperationIndex();
    Assert.assertEquals(new TOID(0, 0, 1), toid);
    toid.incrementOperationIndex();
    Assert.assertEquals(new TOID(0, 0, 2), toid);
    toid.incrementOperationIndex();
    Assert.assertEquals(new TOID(0, 0, 3), toid);

    toid = new TOID(0, 0, 0xFFF);
    toid.incrementOperationIndex();
    Assert.assertEquals(new TOID(1, 0, 0), toid);
    toid.incrementOperationIndex();
    Assert.assertEquals(new TOID(1, 0, 1), toid);
  }

  @Test
  public void testAfterLedger() {
    Assert.assertEquals(4294967295L, TOID.afterLedger(0).toInt64());
    Assert.assertEquals(8589934591L, TOID.afterLedger(1).toInt64());
    Assert.assertEquals(433791696895L, TOID.afterLedger(100).toInt64());
  }

  @Test
  public void initWithInvalidParamsThrows() {
    Assert.assertThrows(IllegalArgumentException.class, () -> new TOID(-1, 0, 0));
    Assert.assertThrows(IllegalArgumentException.class, () -> new TOID(0, -1, 0));
    Assert.assertThrows(IllegalArgumentException.class, () -> new TOID(0, 0, -1));
    Assert.assertThrows(IllegalArgumentException.class, () -> new TOID(0, 0xFFFFF + 1, 0));
    Assert.assertThrows(IllegalArgumentException.class, () -> new TOID(0, 0, 0xFFF + 1));
  }

  @Test
  public void fromInt64WithInvalidParamsThrows() {
    Assert.assertThrows(IllegalArgumentException.class, () -> TOID.fromInt64(-1));
  }

  @Test
  public void testLedgerRangeInclusiveWithInvalidParamsThrows() {
    Assert.assertThrows(IllegalArgumentException.class, () -> TOID.ledgerRangeInclusive(2, 1));
    Assert.assertThrows(IllegalArgumentException.class, () -> TOID.ledgerRangeInclusive(0, 1));
    Assert.assertThrows(IllegalArgumentException.class, () -> TOID.ledgerRangeInclusive(-1, 100));
  }
}

File: src/test/kotlin/org/stellar/sdk/TOIDTest.kt
package org.stellar.sdk

import io.kotest.assertions.throwables.shouldThrow
import io.kotest.core.spec.style.FunSpec
import io.kotest.datatest.withData
import io.kotest.matchers.shouldBe

class TOIDTest :
  FunSpec({
    val ledgerFirst = 1L shl 32
    val txFirst = 1L shl 12
    val opFirst = 1L

    context("test toInt64 and fromInt64") {
      withData(
        nameFn = { (toid, id) ->
          "TOID(${toid.ledgerSequence}, ${toid.transactionOrder}, ${toid.operationIndex}) should be $id"
        },
        TOID(0, 0, 4095) to 4095L,
        TOID(0, 1048575, 0) to 4294963200L,
        TOID(2147483647, 0, 0) to 9223372032559808512L,
        TOID(1, 1, 1) to (ledgerFirst + txFirst + opFirst),
        TOID(1, 1, 0) to (ledgerFirst + txFirst),
        TOID(1, 0, 1) to (ledgerFirst + opFirst),
        TOID(1, 0, 0) to ledgerFirst,
        TOID(0, 1, 0) to txFirst,
        TOID(0, 0, 1) to opFirst,
        TOID(0, 0, 0) to 0L,
        TOID(2147483647, 1048575, 4095) to 9223372036854775807L,
      ) { (toid, id) ->
        toid.toInt64() shouldBe id
        TOID.fromInt64(id) shouldBe toid
      }
    }

    context("test ledgerRangeInclusive") {
      test("ledger 1 to 1") {
        TOID.ledgerRangeInclusive(1, 1) shouldBe
          TOID.TOIDRange(TOID(0, 0, 0).toInt64(), TOID(2, 0, 0).toInt64())
      }
      test("ledger 1 to 2") {
        TOID.ledgerRangeInclusive(1, 2) shouldBe
          TOID.TOIDRange(TOID(0, 0, 0).toInt64(), TOID(3, 0, 0).toInt64())
      }
      test("ledger 2 to 2") {
        TOID.ledgerRangeInclusive(2, 2) shouldBe
          TOID.TOIDRange(TOID(2, 0, 0).toInt64(), TOID(3, 0, 0).toInt64())
      }
      test("ledger 2 to 3") {
        TOID.ledgerRangeInclusive(2, 3) shouldBe
          TOID.TOIDRange(TOID(2, 0, 0).toInt64(), TOID(4, 0, 0).toInt64())
      }
    }

    context("test incrementOperationIndex") {
      test("increment from 0") {
        val toid = TOID(0, 0, 0)
        toid.incrementOperationIndex()
        toid shouldBe TOID(0, 0, 1)
        toid.incrementOperationIndex()
        toid shouldBe TOID(0, 0, 2)
        toid.incrementOperationIndex()
        toid shouldBe TOID(0, 0, 3)
      }

      test("increment from max operation index") {
        val toid = TOID(0, 0, 0xFFF)
        toid.incrementOperationIndex()
        toid shouldBe TOID(1, 0, 0)
        toid.incrementOperationIndex()
        toid shouldBe TOID(1, 0, 1)
      }
    }

    context("test afterLedger") {
      test("after ledger 0") { TOID.afterLedger(0).toInt64() shouldBe 4294967295L }
      test("after ledger 1") { TOID.afterLedger(1).toInt64() shouldBe 8589934591L }
      test("after ledger 100") { TOID.afterLedger(100).toInt64() shouldBe 433791696895L }
    }

    context("test invalid parameters") {
      test("constructor with invalid params throws") {
        shouldThrow<IllegalArgumentException> { TOID(-1, 0, 0) }
        shouldThrow<IllegalArgumentException> { TOID(0, -1, 0) }
        shouldThrow<IllegalArgumentException> { TOID(0, 0, -1) }
        shouldThrow<IllegalArgumentException> { TOID(0, 0xFFFFF + 1, 0) }
        shouldThrow<IllegalArgumentException> { TOID(0, 0, 0xFFF + 1) }
      }

      test("fromInt64 with invalid params throws") {
        shouldThrow<IllegalArgumentException> { TOID.fromInt64(-1) }
      }

      test("ledgerRangeInclusive with invalid params throws") {
        shouldThrow<IllegalArgumentException> { TOID.ledgerRangeInclusive(2, 1) }
        shouldThrow<IllegalArgumentException> { TOID.ledgerRangeInclusive(0, 1) }
        shouldThrow<IllegalArgumentException> { TOID.ledgerRangeInclusive(-1, 100) }
      }
    }
  })
